### DISCUSSION TITLE
Icon Font not rendering correctly

### DISCUSSION THREAD

**Discussion by Trider12** [ocornut/imgui#5275] [New Users Build/Link/Run issues ONLY!]
**_(ADMIN comment added in 2025 --> "see https://github.com/ocornut/imgui/blob/master/docs/FONTS.md")_**

----

**Version/Branch of Dear ImGui:**

Version: Dear ImGui 1.88 WIP (18718)
Branch: master

**Back-end/Renderer/Compiler/OS**

Back-ends: imgui_impl_sdl.cpp + imgui_impl_opengl3.cpp
Compiler: MSVC | Clang 10.0.0
Operating System: Windows 10 | Ubuntu 20.04.4

**My Issue/Question:**

I'm using `IconsFontAwesome5.h` from [IconFontCppHeaders](https://github.com/juliettef/IconFontCppHeaders) and [fa-regular-400.ttf](https://github.com/FortAwesome/Font-Awesome/blob/5.x/webfonts/fa-regular-400.ttf).

**Screenshots/Video**

![image](https://user-images.githubusercontent.com/47952308/166151760-39143fb9-8ea8-4509-ac29-7892880778d3.png)

**Standalone, minimal, complete and verifiable example:**
```cpp
ImGuiIO &io = ImGui::GetIO();
io.Fonts->AddFontDefault();

static const ImWchar icons_ranges[] = {ICON_MIN_FA, ICON_MAX_FA, 0};
ImFontConfig icons_config;
icons_config.MergeMode = true;
icons_config.PixelSnapH = true;

//assert(std::filesystem::exists("assets/fonts/fa-regular-400.ttf"));
io.Fonts->AddFontFromFileTTF("assets/fonts/fa-regular-400.ttf", 16.f, &icons_config, icons_ranges);
```
```cpp
ImGui::SetNextWindowSize({400, 200});
ImGui::Begin("TestWindow");

ImGui::Text(ICON_FA_PAINT_BRUSH " Paint");
ImGui::Button(ICON_FA_SEARCH " Search");

ImGui::End();
```

**Comment by Trider12** [comment]
Using `Metrics/Debugger` I can see that the font does indeed have renderable glyphs:

![image](https://user-images.githubusercontent.com/47952308/166276795-c95822e4-e8ec-4b60-92db-36d0e0ac2b91.png)

**Comment by Trider12** [comment]
So, of course, the issue happened to be with the font itself. The [IconFontCppHeaders example](https://github.com/juliettef/IconFontCppHeaders#example-code) used [fa-solid-900.ttf](https://github.com/FortAwesome/Font-Awesome/blob/6.x/webfonts/fa-solid-900.ttf) and [ImGui example](https://github.com/ocornut/imgui/blob/master/docs/FONTS.md#using-icon-fonts) used [fontawesome-webfont.ttf](https://github.com/FortAwesome/Font-Awesome/blob/4.x/fonts/fontawesome-webfont.ttf) while I tried to use [fa-regular-400.ttf](https://github.com/FortAwesome/Font-Awesome/blob/6.x/webfonts/fa-regular-400.ttf).

The confusion came from these lines in the [IconFontCppHeaders example](https://github.com/juliettef/IconFontCppHeaders#example-code), that do not mention the reduced glyph set of the regular version and then explicitly use the `ICON_FA_PAINT_BRUSH` character which it lacks:
```cpp
io.Fonts->AddFontFromFileTTF( FONT_ICON_FILE_NAME_FAS, 16.0f, &icons_config, icons_ranges );
// use FONT_ICON_FILE_NAME_FAR if you want regular instead of solid

// in an imgui window somewhere...
ImGui::Text( ICON_FA_PAINT_BRUSH "  Paint" );    // use string literal concatenation
```

Switching to [fontawesome-webfont.ttf](https://github.com/FortAwesome/Font-Awesome/blob/4.x/fonts/fontawesome-webfont.ttf) or [forkawesome-webfont.ttf](https://github.com/ForkAwesome/Fork-Awesome/blob/master/fonts/forkawesome-webfont.ttf) fixed the issue:

![image](https://user-images.githubusercontent.com/47952308/166297000-7c893502-0873-438e-8704-7d59c8ec496d.png)

**Reply by svkvarun43** [reply]
Hi,

I followed your steps and have the header `IconsFontAwesome5.h` and the `fontawesome-webfont.ttf` as well as `forkawesome-webfont.ttf`

```
    #include "IconsFontAwesome5.h"
    
    ImGuiIO &io = ImGui::GetIO();
    io.Fonts->AddFontDefault();
    
    static const ImWchar icons_ranges[] = {ICON_MIN_FA, ICON_MAX_FA, 0};
    ImFontConfig icons_config;
    icons_config.MergeMode = true;
    icons_config.PixelSnapH = true;
    
    io.Fonts->AddFontFromFileTTF("fontawesome-webfont.ttf", 16.0f, &icons_config, icons_ranges);
```

`ImGui::Text( ICON_FA_PAINT_BRUSH "  Paint" );`

I still get ? Paint in my output.

![image](https://user-images.githubusercontent.com/98422353/172697586-312464b7-8a22-40ef-9bc6-5654cb3f65e3.png)

The `.tff` files are in the same location as my xxx.cpp file 

Could you please help?



Thanks

**Reply by ocornut** [reply]
We now have variety of tools to help people debug their issue.

1. Call `ImGui::DebugTextEncoding(ICON_FA_PAINT_BRUSH);` to see with codepoints are sent in your text.
2. Browse `Metrics->Fonts` to confirm which glyphs are loaded.

**Reply by svkvarun43** [reply]
Thanks for the reply,

I am unfortunately bound by the imgui version in this fork [https://github.com/benui-dev/UnrealImGui](https://github.com/benui-dev/UnrealImGui) which doesn't have DebugTextEncoding as im using it as a plugin for Unreal Engine 5.0. Is there any other way to diagnose the issue please?

**Reply by Trider12** [reply]
@svkvarun43, you are using the wrong header file:
* `IconsFontAwesome5.h` is from the version 5.x
* [fontawesome-webfont.ttf](https://github.com/FortAwesome/Font-Awesome/blob/4.x/fonts/fontawesome-webfont.ttf) is from the version 4.x
* [forkawesome-webfont.ttf](https://github.com/ForkAwesome/Fork-Awesome/blob/master/fonts/forkawesome-webfont.ttf) is from the fork version

You need to use an appropriate header for each `.ttf`.

**Reply by svkvarun43** [reply]
@Trider12 

Thanks for the reply

So I replaced the header now to `#include "IconsFontAwesome4.h"` with [fontawesome-webfont.ttf](https://github.com/FortAwesome/Font-Awesome/blob/4.x/fonts/fontawesome-webfont.ttf), didn't work

and also tried with ` #include IconFontAwesome6.h` with its [fa-solid-900.ttf](https://github.com/FortAwesome/Font-Awesome/blob/6.x/webfonts/fa-solid-900.ttf). 

But I still end up with a ? for any icon i try to load :(

**Reply by Trider12** [reply]
@svkvarun43, as @ocornut have stated above, use the `Metrics` window to inspect the glyphs.

**Reply by blank-repo** [reply]
@Trider12 
Thank you!
I just wanted basic emojis I don't even care which.
I was running around a bunch of emoji-related issues here with not even a hint,
until you pointed to [fontawesome-webfont.ttf](https://github.com/FortAwesome/Font-Awesome/blob/4.x/fonts/fontawesome-webfont.ttf)
solved for me 👍 
Thank you sir/ma'am.

**Comment by ocornut** [comment]
You can update the copy of dear imgui in that repo.
